enclave-tls/cmake/SGXCommon.cmake

```diff
@@ -0,0 +1,313 @@
+        set(TLIB_LIST "${TLIB_LIST} -l${TLIB}")
+    endforeach()
+
+    set_target_properties(${target} PROPERTIES COMPILE_FLAGS ${ENCLAVE_C_FLAGS})
+    target_include_directories(${target} PRIVATE ${CMAKE_CURRENT_BINARY_DIR} ${ENCLAVE_INCLUDES})
+    set_property(DIRECTORY APPEND PROPERTY ADDITIONAL_MAKE_CLEAN_FILES "${EDL_T_H} ${EDL_T_C}")
+endfunction()
+
+# Build enclave shared library
+function(add_enclave_library target)
+    set(optionArgs USE_PREFIX)
+    set(oneValueArgs EDL_OBJ LDSCRIPT)
+    set(multiValueArgs SRCS TRUSTED_LIBS EDL EDL_SEARCH_PATHS)
+    cmake_parse_arguments("SGX" "${optionArgs}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+    if("${SGX_EDL}" STREQUAL "")
+        message("${target}: SGX enclave edl file is not provided!")
+    endif()
+
+    if("${SGX_EDL_SEARCH_PATHS}" STREQUAL "")
+        message("${target}: SGX enclave edl file search paths are not provided!")
+    endif()
+
+    if(NOT "${SGX_LDSCRIPT}" STREQUAL "")
+        get_filename_component(LDS_ABSPATH ${SGX_LDSCRIPT} ABSOLUTE)
+        set(LDSCRIPT_FLAG "-Wl,--version-script=${LDS_ABSPATH}")
+    endif()
+
+    if("${SGX_EDL}" STREQUAL "" AND "${SGX_EDL_OBJ}" STREQUAL "")
+        add_library(${target} SHARED ${SGX_SRCS})
+    elseif(NOT "${SGX_EDL_OBJ}" STREQUAL "")
+        add_library(${target} SHARED ${SGX_SRCS} $<TARGET_OBJECTS:${SGX_EDL_OBJ}>)
+    elseif(NOT "${SGX_EDL}" STREQUAL "")
+        if(${SGX_USE_PREFIX})
+            build_edl_trust(${target} EDL ${SGX_EDL} EDL_SEARCH_PATHS ${SGX_EDL_SEARCH_PATHS} USE_PREFIX ${SGX_USE_PREFIX})
+        else()
+            build_edl_trust(${target} EDL ${SGX_EDL} EDL_SEARCH_PATHS ${SGX_EDL_SEARCH_PATHS})
+        endif()
+        add_library(${target} SHARED ${SGX_SRCS} $<TARGET_OBJECTS:${target}-edlobj>)
+    endif()
+
+    set(TLIB_LIST "")
+    foreach(TLIB ${SGX_TRUSTED_LIBS})
+        set(TLIB_LIST "${TLIB_LIST} -l${TLIB}")
+    endforeach()
+    set(ENCLAVE_LINK_FLAGS "-m64 -Wall -Wl,--no-undefined -nostdlib -nodefaultlibs -nostartfiles -L${SGX_LIBRARY_PATH}")
+    set(ENCLAVE_LINK_FLAGS "${ENCLAVE_LINK_FLAGS} -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -pie")
+    set(ENCLAVE_LINK_FLAGS "${ENCLAVE_LINK_FLAGS} -Wl,--whole-archive ${TLIB_LIST} -l${SGX_TRTS_LIB} -Wl,--no-whole-archive")
+    set(ENCLAVE_LINK_FLAGS "${ENCLAVE_LINK_FLAGS} -Wl,--whole-archive -lsgx_tsgxssl -Wl,--no-whole-archive")
+    set(ENCLAVE_LINK_FLAGS "${ENCLAVE_LINK_FLAGS} -Wl,--start-group -lsgx_tstdc -lsgx_pthread -lsgx_tcxx -lsgx_tkey_exchange -lsgx_tcrypto -lsgx_trts -lsgx_tservice -lsgx_tsgxssl_crypto -lsgx_tsgxssl_ssl -l${SGX_TSVC_LIB} -l${SGX_DCAP_TVL} -Wl,--end-group")
+    set(ENCLAVE_LINK_FLAGS "${ENCLAVE_LINK_FLAGS} -Wl,-Bstatic -Wl,-Bsymbolic -Wl,--no-undefined")
+    set(ENCLAVE_LINK_FLAGS "${ENCLAVE_LINK_FLAGS} -Wl,-pie,-eenclave_entry -Wl,--export-dynamic")
+    set(ENCLAVE_LINK_FLAGS "${ENCLAVE_LINK_FLAGS} ${LDSCRIPT_FLAG}")
+    set(ENCLAVE_LINK_FLAGS "${ENCLAVE_LINK_FLAGS} -Wl,--defsym,__ImageBase=0")
```

Comment:
Is it better to pass these link options as this common function's arguments?

Reply:
it's the same as current implementation.

Reply:
Yes, I know it's Yun Ya's implementation. But I think as this is a common function, so these speific options should be passed as augments.

Reply:
what's different between passing argment and this kind of implementation?